File: backend/controllers/userController.js
import User from '../models/User.js';
import cloudinary from '../config/cloudinary.js';
import multer from 'multer';

//memory storage so we can upload directly to Cloudinary
const storage = multer.memoryStorage();
export const upload = multer({ 
  storage,
  limits: { fileSize: 5 * 1024 * 1024 }, 
  fileFilter: (req, file, cb) => {
    // Only allow image files
    if (file.mimetype.startsWith('image/')) {
      cb(null, true);
    } else {
      cb(new Error('Only image files are allowed'), false);
    }
  }
});

export const getProfile = async (req, res) => {
  try {
    const user = await User.findById(req.user.id).select('-password');
    if (!user) return res.status(404).json({ msg: 'User not found' });
    res.json(user);
  } catch (err) {
    console.error('Get profile error:', err);
    res.status(500).json({ msg: 'Server error' });
  }
};


export const updateProfile = async (req, res) => {
  try {
    // Define allowed fields to prevent unauthorized updates
    const allowedUpdates = ['name', 'email', 'bio', 'location']; // Adjust as needed
    const updates = {};

    Object.keys(req.body).forEach(key => {
      if (allowedUpdates.includes(key)) {
        updates[key] = req.body[key];
      }
    });


    if (Object.keys(updates).length === 0) {
      return res.status(400).json({ msg: 'No valid fields to update' });
    }

    const user = await User.findByIdAndUpdate(
      req.user.id, 
      updates, 
      { new: true, runValidators: true }
    ).select('-password');
    
    if (!user) return res.status(404).json({ msg: 'User not found' });
    
    res.json(user);
  } catch (err) {
    console.error('Update profile error:', err);
    if (err.name === 'ValidationError') {
      return res.status(400).json({ msg: 'Validation error', error: err.message });
    }
    res.status(500).json({ msg: 'Server error' });
  }
};


export const uploadAvatar = [
  upload.single('avatar'),
  async (req, res) => {
    try {
      if (!req.file) {
        return res.status(400).json({ msg: 'No file uploaded' });
      }

      const uploadToCloudinary = () => {
        return new Promise((resolve, reject) => {
          const uploadStream = cloudinary.uploader.upload_stream(
            {
              folder: 'avatars',
              resource_type: 'image',
              transformation: [
                { width: 400, height: 400, crop: 'fill' },
                { quality: 'auto' }
              ]
            },
            (error, result) => {
              if (error) {
                reject(error);
              } else {
                resolve(result);
              }
            }
          );
          

          uploadStream.end(req.file.buffer);
        });
      };


      const result = await uploadToCloudinary();
      
      const user = await User.findByIdAndUpdate(
        req.user.id,
        { avatar: result.secure_url },
        { new: true }
      ).select('-password');

      if (!user) {
        return res.status(404).json({ msg: 'User not found' });
      }

      res.json({
        msg: 'Avatar uploaded successfully',
        user,
        avatarUrl: result.secure_url
      });

    } catch (err) {
      console.error('Upload avatar error:', err);
      if (err.message === 'Only image files are allowed') {
        return res.status(400).json({ msg: 'Only image files are allowed' });
      }
      res.status(500).json({ msg: 'Server error', error: err.message });
    }
  }
];

File: backend-flask/controllers/user_controller.py
from flask import request, jsonify
from models.user import User
from middlewares.auth_middleware import auth_required
import cloudinary
import cloudinary.uploader

@auth_required
def get_profile():
    """Get user profile"""
    try:
        user = User.find_by_id(request.user['id'])
        if not user:
            return jsonify({'msg': 'User not found'}), 404
        
        # Remove password from response
        user_data = {k: v for k, v in user.items() if k != 'password'}
        return jsonify(user_data), 200
    
    except Exception as e:
        print(f"Get profile error: {str(e)}")
        return jsonify({'msg': 'Server error', 'error': str(e)}), 500

@auth_required
def upload_avatar():
    """Upload user avatar to Cloudinary"""
    try:
        if 'avatar' not in request.files:
            return jsonify({'msg': 'No file uploaded'}), 400
        
        file = request.files['avatar']
        
        # Check if file is empty
        if file.filename == '':
            return jsonify({'msg': 'No file selected'}), 400
        
        # Validate file type
        if not file.content_type.startswith('image/'):
            return jsonify({'msg': 'Only image files are allowed'}), 400
        
        print(f"📤 Uploading avatar for user {request.user['id']}")
        
        # Upload to Cloudinary
        result = cloudinary.uploader.upload(
            file,
            folder='avatars',
            transformation=[
                {'width': 400, 'height': 400, 'crop': 'fill'},
                {'quality': 'auto'}
            ]
        )
        
        print(f"✅ Upload successful: {result['secure_url']}")
        
        # Update user
        user = User.update(request.user['id'], avatar=result['secure_url'])
        user_data = {k: v for k, v in user.items() if k != 'password'}
        
        return jsonify({
            'msg': 'Avatar uploaded successfully',
            'user': user_data,
            'avatarUrl': result['secure_url']
        }), 200
    
    except Exception as e:
        print(f"❌ Upload avatar error: {str(e)}")
        import traceback
        traceback.print_exc()
        return jsonify({'msg': 'Server error', 'error': str(e)}), 500